sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java

```diff
@@ -167,9 +167,21 @@ public boolean equals(Object other) {
         return false;
       }
       for (int i = 0; i < n; i ++) {
-        if (isNullAt(i) != row.isNullAt(i) || (!isNullAt(i) && !get(i).equals(row.get(i)))) {
+        if (isNullAt(i) != row.isNullAt(i)) {
           return false;
         }
+        if (!isNullAt(i)) {
+          Object o1 = get(i);
+          Object o2 = row.get(i);
+          if (o1 instanceof byte[]) {
+            // handle equals() of byte[]
+            if (!(o2 instanceof byte[]) || !java.util.Arrays.equals((byte[])o1, (byte[])o2)) {
+              return false;
+            }
+          } else if (!o1.equals(o2)) {
+            return false;
+          }
+        }
       }
       return true;
```

Comment:
Because `Row` is a trait, so we can't use it in Java (for UnsafeRow and SpecificRow in codegen), so we at least need two version of it.